test/lib/jdk/test/lib/process/OutputAnalyzer.java

```diff
@@ -786,4 +794,59 @@ private int indexOf(List<String> lines, String regexp, int fromIndex) {
         return -1;
     }
 
+    private void searchLinesForMultiLinePattern(String[] haystack, String[] needles, boolean verbose) {
+        int needleIdx = 0;
+        int haystackIdx = 0;
+        boolean started = false;
+        boolean found = false;
+        while (needleIdx < needles.length && haystackIdx < haystack.length && (started == false || found == true)) {
+            if (verbose) {
+                System.out.println("" + haystackIdx + ":" + haystack[haystackIdx]);
+            }
+            found = haystack[haystackIdx].contains(needles[needleIdx]);
+            if (found) {
+                if (verbose) {
+                    System.out.println("Matches pattern " + needleIdx + " (\"" + haystack[haystackIdx] + "\")");
+                }
+                started = true;
+                needleIdx++;
+            }
+            haystackIdx++;
+        }
+        if (needleIdx < needles.length) {
+            String err = "First unmatched: \"" + needles[needleIdx] + "\"";
+            if (!verbose) { // don't print twice
+                reportDiagnosticSummary();
+            }
+            throw new RuntimeException(err);
+        }
```

Comment:
Splitting this up into multiple loops would simplify the code significantly imho, no need for two boolean flags for a start. Also, what does the `!verbose` check do exactly? Doesn't seem like it would be print twice without it, to me.

```java
if (needles.length == 0) return;

int haystackStart = 0;
for (int i = 0; i < haystack.length; i++) {
     if (haystack[i].contains(needles[0])) {
      hayStackStart = i+1;
      return;
    }
}

for (int i = 1; i < needles.length; i++) {
    if (verbose && haystackStart + i < haystack.length) {
        System.out.println("" + haystackStart + i + ":" + haystack[haystackStart + i]);
    }
    if (haystackStart + i >= haystack.length
        || !haystack[haystackStart + i].contains(needles[i])) {
        String err = "First unmatched: \"" + needles[needleIdx] + "\"";
        if (!verbose) { // don't print twice
            reportDiagnosticSummary();
        }
        throw new RuntimeException(err);
    }
    if (verbose) {
        System.out.println("Matches pattern " + needleIdx + " (\"" + haystack[haystackStart + i] + "\")");
    }
}
```
```

Reply:
> Splitting this up into multiple loops would simplify the code significantly imho, no need for two boolean flags for a start. 

If verbose=true, we print the haystack, with pattern matches marked. Your version omits printing for all lines until a needle is found, and only prints from the second needle on. Meaning, if the first pattern does not match, we don't see output. Arguably, that is the point where you need the output for test analysis.

Okay, I can give it a try.

> Also, what does the !verbose check do exactly? Doesn't seem like it would be print twice without it, to me.

If an error happened, we print diagnostic summary (which mimics the behavior of all other "shouldBlaBla()" functions). So, we print twice if we run with verbose and run into an error.

Reply:
Okay, I rewrote it in the style you prefer, while keeping the logs I wanted intact; this does not look much clearer to me tbh. I found the first version more elegant, since it was a one-loop state machine and could be easily adapted to other uses (e.g. scanning for pattern that are not in directly consecutive lines, but interspersed)

No matter. I got it to work, lets keep it this way.

Reply:
>If an error happened, we print diagnostic summary (which mimics the behavior of all other "shouldBlaBla()" functions). So, we print twice if we run with verbose and run into an error.

Thank you for explaining!

>I found the first version more elegant, since it was a one-loop state machine and could be easily adapted to other uses (e.g. scanning for pattern that are not in directly consecutive lines, but interspersed)

That's just differing taste then, thank you for taking the effort to try out my idea :-).